aten/src/ATen/native/cuda/Math.cuh

```diff
@@ -196,5 +196,144 @@ static inline C10_HOST_DEVICE scalar_t calc_gcd(scalar_t a_in, scalar_t b_in) {
   return b;
 }
 
+/*
+ * The following function was converted to CUDA form from code that comes
+ * with the following copyright notice. It has been released under the BSD license.
+ *
+ * Cephes Math Library Release 2.0:  April, 1987
+ * Copyright 1985, 1987 by Stephen L. Moshier
+ * Direct inquiries to 30 Frost Street, Cambridge, MA 02140
+ *
+ * Evaluates the series
+ *
+ *       len-1
+ *         - '
+ *  y  =   >   array[i] T (x/2)
+ *         -             i
+ *        i=0
+ *
+ * of Chebyshev polynomials Ti at argument x/2.
+ *
+ * Coefficients are stored in reverse order, i.e. the zero order term is last in the array.  Note len is the number of
+ * coefficients, not the order.
+ *
+ * If coefficients are for the interval a to b, x must have been transformed to x -> 2(2x - b - a)/(b-a) before
+ * entering the routine.  This maps x from (a, b) to (-1, 1), over which the Chebyshev polynomials are defined.
+ *
+ * If the coefficients are for the inverted interval, in which (a, b) is mapped to (1/b, 1/a), the transformation
+ * required is x -> 2(2ab/x - b - a)/(b-a).  If b is infinity, this becomes x -> 4a/x - 1.
+ */
+template <typename scalar_t>
+static inline C10_HOST_DEVICE scalar_t chbevl(scalar_t _x, const scalar_t array[], size_t len) {
+  using accscalar_t = at::acc_type<scalar_t, true>;
+  accscalar_t x = static_cast<accscalar_t>(_x);
+  accscalar_t b0, b1, b2;
+
+  b0 = static_cast<accscalar_t>(array[0]);
+  b1 = 0;
+
+  for (size_t i = 1; i < len; ++i)  {
+    b2 = b1;
+    b1 = b0;
+    b0 = x * b1 - b2 + static_cast<accscalar_t>(array[i]);
+  }
+
+  return static_cast<scalar_t>(0.5 * (b0 - b2));
+}
+
+/*
+ * The following function has been modified from code that comes with the following copyright notice.
+ *
+ * Cephes Math Library Release 2.8:  June, 2000
+ * Copyright 1984, 1987, 2000 by Stephen L. Moshier
+ *
+ * Returns modified Bessel function of order zero of the argument.
+ * The range is partitioned into the two intervals [0,8] and (8, infinity). Chebyshev polynomial expansions are employed
+ * in each interval.
+ */
+template <typename scalar_t>
+static inline C10_HOST_DEVICE scalar_t calc_i0(scalar_t _x) {
+  using accscalar_t = at::acc_type<scalar_t, true>;
+  accscalar_t x = ::abs(static_cast<accscalar_t>(_x));
```

Comment:
This will perform the computation in float32 with bfloat16 and float16 inputs, right?

That may be the correct thing to do, but before jumping to that conclusion we should look at the numerical accuracy (or inaccuracy) of computing in bfloat16/float16 first.

Reply:
Update: Same comment as above explaining the upcasting would go here, too